module/zfs/spa_errlog.c

```diff
@@ -128,6 +241,300 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
 	mutex_exit(&spa->spa_errlist_lock);
 }
 
+#ifdef _KERNEL
+static int
+find_block_txg(dsl_dataset_t *ds, zbookmark_err_phys_t *zep,
+    uint64_t *birth_txg)
+{
+	objset_t *os;
+	int error = dmu_objset_from_ds(ds, &os);
+	if (error != 0)
+		return (error);
+
+	dnode_t *dn;
+	blkptr_t bp;
+
+	error = dnode_hold(os, zep->zb_object, FTAG, &dn);
+	if (error != 0)
+		return (error);
+
+	rw_enter(&dn->dn_struct_rwlock, RW_READER);
+	error = dbuf_dnode_findbp(dn, zep->zb_level, zep->zb_blkid, &bp, NULL,
+	    NULL);
+
+	if (error == 0 && BP_IS_HOLE(&bp))
+		error = SET_ERROR(ENOENT);
+
+	*birth_txg = bp.blk_birth;
+	rw_exit(&dn->dn_struct_rwlock);
+	dnode_rele(dn, FTAG);
+	return (error);
+}
+
+/*
+ * This function serves a double role. If only_count is true, it returns
+ * how many times an error block belonging to this filesystem is referenced by
+ * snapshots or clones. If only_count is false, each time the error block is
+ * referenced by a snapshot or clone, it returns a pointer to the bookmark of
+ * the error block back to process_error_log().
```

Comment:
I don't see a pointer being returned.  It looks like are filling the userspace array at `addr` with `zbookmark_phys_t`'s.  The array has space for `*count` entries (each entry holds a `zbookmark_phys_t`).  The array is filled from the back, and the *count is modified to be the number of unused entries at the beginning of the array.

This seems like an odd interface.  Maybe it's following an existing precedent?  In which case we should at least document it well.  If not, could we instead fill the array from the beginning, and return the number of entries added?

Reply:
I realized my wording is wrong. It does not return a pointer, it copies to userspace (in *addr, or *uaddr if you prefer) the bookmark of the error block. This is a pre-existing interface, it was not implemented in this PR.